FilmlisteParser.js

```diff
@@ -1,138 +1,160 @@
-                                url_website: parsed[9],
-                                url_subtitle: parsed[10],
-                                url_video: parsed[8],
-                                url_video_low: createUrlFromBase(parsed[8], parsed[12]),
-                                url_video_hd: createUrlFromBase(parsed[8], parsed[14])
-                            };
-
-                            buffer.push(['sadd', setKey, JSON.stringify(entry)]);
-                            entries++;
-
-                            if (currentLine % 500 == 0) {
-                                redis.batch(buffer).exec();
-                                buffer = [];
-                                ipc.send('state', {
-                                    entries: entries,
-                                    progress: getProgress()
-                                });
-                            }
-                        } else if (last) {
-                            redis.batch(buffer).exec(() => {
-                                ipc.send('state', {
-                                    entries: entries,
-                                    progress: 1
-                                });
-                                ipc.send('done');
-
-                                fs.close(fd, () => {
-                                    setTimeout(() => process.exit(0), 500);
-                                });
-                            });
-                        }
-
-                        getNext();
-                    });
-                }
+  fs.open(file, 'r', (err, fd) => {
+    if (err) {
+      handleError(err);
+      return;
+    }
+
+    fs.fstat(fd, (err, stats) => {
+      if (err) {
+        handleError(err);
+        return;
+      }
+
+      let currentChannel,
+          currentTopic,
+          buffer = [],
+          currentLine = 0;
+
+      const filesize = stats.size,
+            fileStream = fs.createReadStream(null, {fd: fd, autoClose: true}),
+            lineReaderSeparator = {separator: /^{"Filmliste":|,"X":|}$/}
+
+      const getProgress = () => {
+        return fileStream.bytesRead / filesize;
+      };
+
+      lineReader.eachLine(fileStream, lineReaderSeparator, (line, last, getNext) => {
+        currentLine++;
+        if (currentLine === 1) {
+          return getNext();
+        }
+        if (currentLine === 2) {
+          redis.set(timestampKey, handleListMeta({line}));
+          return getNext();
+        }
+
+        if (last) {
+          return redis.batch(buffer).exec(() => {
+            ipc.send('state', {
+              entries: currentLine - 2,
+              progress: 1
             });
+            ipc.send('done');
+            fs.close(fd, () => setTimeout(() => process.exit(0), 500));
+          });
         }
-    });
 
+        [currentChannel, currentTopic, entry] = mapListLineToRedis({line, currentChannel, currentTopic})
```

Comment:
I didn't defined them newly every time, I updated them whenever the filmliste has non 0-length entries.

I also don't get why it isn't real json, but I can't do anything about it, at least not yet. In long term I plan to create a new crawler, which has a "moving" filmliste in a real database by crawling all the time (throttled) and not once every 2 hours, besides some other things. But I haven't stated that idea to the MV-team yet.